lib/experimental/class-wp-html-processor.php

```diff
@@ -0,0 +1,720 @@
+		} elseif ( is_array( $data_attribute ) && is_string( $data_attribute[0] ) && ( is_callable( $data_attribute[1] ) || function_exists( $data_attribute[1] ) ) ) {
+			$descriptor->data_attribute = $data_attribute;
+		} else {
+			$descriptor->data_attribute = null;
+		}
+
+		return $descriptor;
+	}
+
+
+	public static function comparable( $input ) {
+		return strtolower( trim( $input ) );
+	}
+
+	/**
+	 * @param string $tag
+	 * @param array<WP_HTML_Attribute_Token> $attributes
+	 * @return 'matches'|'cannot-match'|'might-match'|'not-implemented'
+	 */
+	public function check( $tag, $attributes ) {
+		// @TODO Special-case h1..h6 here
+		if ( $this->tag_name !== null && $this->tag_name !== $tag ) {
+			return 'cannot-match';
+		}
+
+		if ( null === $this->class_name && null === $this->data_attribute ) {
+			return 'matches';
+		}
+
+		if ( null === $this->data_attribute && isset( $attributes['class'] ) ) {
+			// @TODO: Avoid this nested if
+			if ( null === $attributes['class']->value ) {
+				return false;
+			}
+
+			$pattern_class = preg_quote( $this->class_name, '~' );
+			return 1 === preg_match( "~(?:^|[\t ]){$pattern_class}(?:[\t ]|$)~Smui", $attributes['class']->value->comparable );
+		}
+
+		if ( null === $this->class_name ) {
+			return 'not-implemented';
+		}
+
+		return 'might-match';
+	}
+}
+
+
+class WP_HTML_Scanner_Token {
+	/**
+	 * The starting byte offset in some document where the token was found.
+	 *
+	 * @var int
+	 */
+	public $start;
+
+	/**
+	 * The byte length of the token in the source document.
+	 *
+	 * @var int
+	 */
+	public $length;
+
+	/**
+	 * Comparable value of the token
+	 *
+	 * @var string|null
+	 */
+	public $comparable = null;
+
+	public function __construct( $start, $length, $comparable ) {
+		$this->start = $start;
+		$this->length = $length;
+		$this->comparable = $comparable;
+	}
+
+	/**
+	 * Converts a `preg_match()` result with PREG_OFFSET_CAPTURE
+	 * into a token pair of start-offset and length.
+	 *
+	 * This class exists to disambiguate alternative ways we could
+	 * be representing a pattern match, and it holds a lazy
+	 * reference to a comparable value to the parsed token.
+	 *
+	 * @param $preg_match
+	 * @return self
+	 */
+	public static function from_preg_match( $preg_match ) {
+		list( $source_value, $starts_at ) = $preg_match;
+
+		return new self( $starts_at, strlen( $source_value ), WP_Tag_Find_Descriptor::comparable( $source_value ) );
+	}
+}
+
+
+class WP_HTML_Attribute_Token {
+	/**
+	 * @var WP_HTML_Scanner_Token
+	 */
+	public $name;
+
+	/**
+	 * @var WP_HTML_Scanner_Token|null
+	 */
+	public $value;
+
+	/**
+	 * @var int
+	 */
+	public $attribute_start;
+
+	/**
+	 * @var int
+	 */
+	public $attribute_end;
+
+	public function __construct( $name, $value, $start, $end ) {
+		$this->name            = $name;
+		$this->value           = $value;
+		$this->attribute_start = $start;
+		$this->attribute_end   = $end;
+	}
+}
+
+
+class WP_HTML_Scanner {
```

Comment:
I'd merge this and WP_HTML_Processor into a single class.

True, scanning is conceptually a different operation than updating the HTML. However, they are related closely enough to require public access to the properties of this class.

A single class = minimal public API. And if we need a separate Scanner one day, it would be easy to carve one out from the monolith.